### PR TITLE
Update the EBO static_assert to dheck against APFloat::Storage instead of IEEEFloat.

### PR DESCRIPTION
**Pull request by DanielCChen**
The addition of the the following code in commit https://github.com/llvm/llvm-project/commit/6004f5550c8032f4c632cdbf5dbc0894bb33e51f#diff-3ec11da67f69353fd755e59cc71f551b5e93773fd4e6f6327c7f1c7074a82709R1474

```
static_assert(sizeof(APFloat) == sizeof(detail::IEEEFloat),
              "Empty base class optimization is not performed.");
 ```
              
broke our downstream code as not all the members of APFloat is IEEE.

This PR supersedes PR #111780. 

We made some changed to our downstream code to accommodate the upstream change. 
This PR is to modify the EBO static_assert to check against the `sizeof(APFloat::Storage)` instead of `sizeof(detail::IEEEFloat)`, which seems satisfy both upstream and downstream code.  
We also moved the assert to `APFloat.cpp` from the header file as per discussion in PR #111780 

@Ariel-Burton

### CODE REVIEW

**Comment by arsenm** [comment]
> broke our downstream code as not all the members of APFloat is IEEE.

So I still see no reason to update the upstream code to handle this. Your downstream code is different and should change the assert there

File: llvm/lib/Support/APFloat.cpp [hunk 1]
```diff
@@ -99,6 +99,11 @@ enum class fltNanEncoding {
   NegativeZero,
 };
 
+class APFloatEBOChecker {
```
Comment (kuhar) [inline llvm/lib/Support/APFloat.cpp:102]: You don't have to add a class for this assert -- it should work in the global scope.

Comment (DanielCChen) [reply llvm/lib/Support/APFloat.cpp:102]: The reason is that `APFloat::U` is private, so we added a `friend` class to access it.

**Comment by DanielCChen** [comment]
> > broke our downstream code as not all the members of APFloat is IEEE.
> 
> So I still see no reason to update the upstream code to handle this. Your downstream code is different and should change the assert there

Thanks for the comments. I totally understand your point.
In fact, we are in the process of upstreaming our downstream code as @Ariel-Burton stated in the "old" PR.
I think the question is if the change in this PR satisfies the purpose of the original assertion in PR #111641? If not, we can definitely look into it.

File: llvm/lib/Support/APFloat.cpp (hunk 1 above)
Comment (kuhar) [reply llvm/lib/Support/APFloat.cpp:102]: OK, that's fair. Could you add a comment explaining this next to the friend declaration?

Comment (DanielCChen) [reply llvm/lib/Support/APFloat.cpp:102]: Sure. Will do.

**Comment by DanielCChen** [comment]
Reviewers, could you please take some time to review this PR? If no further comments or objections, I would like to merge it early next week. Thanks in advance!

**Review by nikic — CHANGES_REQUESTED** [review]
I would like to keep the assertion as-is. It found a legitimate issue with your downstream changes: That they increase the size of APFloat.

You should drop the assertion downstream for now. When you upstream your changes, we can consider whether we are willing to take that size increase or not, and adjust the assertion accordingly at that time.

**Comment by DanielCChen** [comment]
> I would like to keep the assertion as-is. It found a legitimate issue with your downstream changes: That they increase the size of APFloat.
> 
> You should drop the assertion downstream for now. When you upstream your changes, we can consider whether we are willing to take that size increase or not, and adjust the assertion accordingly at that time.

Ok. That is fair. We will bring it back once we are ready to upstream our code in the future. 
Thanks again to everyone for reviewing it.

**Pull request closed without merging**